File: java/solutions/src/main/java/io/github/ajoz/workshop/fp/part1/practice3/Practice3.java
package io.github.ajoz.workshop.fp.part1.practice3;

import io.github.ajoz.workshop.fp.tools.Function1;
import io.github.ajoz.workshop.fp.tools.Function2;
import io.github.ajoz.workshop.fp.tools.Predicate;
import io.github.ajoz.workshop.fp.tools.predicates.IntPredicates;
import io.github.ajoz.workshop.fp.tools.predicates.Predicates;

import java.util.ArrayList;
import java.util.List;

class Lists {
    static <A> List<A> select(final Predicate<A> predicate,
                              final List<A> elements) {
        final List<A> selected = new ArrayList<>();
        for (final A element : elements) {
            if (predicate.test(element))
                selected.add(element);
        }
        return selected;
    }

    static <A> List<A> reject(final Predicate<A> predicate,
                              final List<A> elements) {
        return Lists.select(predicate.not(), elements);
    }
}

class ExercisesWithAList {
    static List<Integer> allAbove42(final List<Integer> values) {
        return Lists.select(IntPredicates.isLargerThen(42), values);
    }

    static Function1<List<Integer>, List<Integer>> allBelow42 =
            /*
               Unfortunately javac cannot correctly infer the types of the
               expression and we need to help him by marking the types for the
               Function2.curry method.
              */
            Function2.<Predicate<Integer>, List<Integer>, List<Integer>>curry(Lists::select).apply(IntPredicates.isLowerThen(42));
    // it does not look good or appealing :-( such is the cruel fate of
    // point free FP in Java

}

class Practice3 {
    private static Predicate<String> hasFirstCapitalLetter =
            str -> Character.isUpperCase(str.charAt(0));

    private static Predicate<String> hasLastCapitalLetter =
            str -> Character.isUpperCase(str.charAt(str.length() - 1));

    private static Predicate<String> hasLengthLargerThen(final int length) {
        return str -> str.length() > length;
    }

    private static Predicate<String> hasLenghtSmallerThen(final int length) {
        return str -> str.length() < length;
    }

    private static Predicate<String> contains(final String text) {
        return str -> str.contains(text);
    }

    private static Predicate<String> hasLengthInRange =
            hasLengthLargerThen(3).and(hasLenghtSmallerThen(10));

    private static Predicate<String> isNotNull = Predicates.isNotNull();

    private static Predicate<String> hasCorrectFirstOrLastLetter =
            hasFirstCapitalLetter.or(hasLastCapitalLetter);

    static List<String> getStrings(final List<String> data) {
        return Lists.select(
                isNotNull.and(hasLengthInRange).and(hasCorrectFirstOrLastLetter).and(contains("JUG")),
                data
        );
    }
}

File: kotlin/solutions/src/main/kotlin/org/artrev/workshop/fp/part1/practice3/Practice3.kt
@file:Suppress("unused", "UNUSED_PARAMETER")

package org.artrev.workshop.fp.part1.practice3

import org.artrev.workshop.fp.part1.practice1.not
import org.artrev.workshop.fp.tools.and
import org.artrev.workshop.fp.tools.curry
import org.artrev.workshop.fp.tools.or
import org.artrev.workshop.fp.tools.predicates.isLargerThen
import org.artrev.workshop.fp.tools.predicates.isLowerThen

internal object Lists {
    // Part 1:
    fun <A> select(predicate: (A) -> Boolean, elements: List<A>): List<A> {
        val selected = mutableListOf<A>()
        for (element in elements) {
            if (predicate(element))
                selected.add(element)
        }
        return selected
    }

    // Part 2:
    fun <A> reject(predicate: (A) -> Boolean, elements: List<A>) =
            select(predicate.not(), elements)
}

// Part 3:
fun allAbove42(values: List<Int>): List<Int> =
        Lists.select(isLargerThen(42), values)

// Part 4:
var allBelow42: (List<Int>) -> List<Int> =
        curry<(Int) -> Boolean, List<Int>, List<Int>>(Lists::select)(isLowerThen(42))

// Part 5:
private val hasFirstCapitalLetter: (String) -> Boolean =
        { it.first().isUpperCase() }

private val hasLastCapitalLetter: (String) -> Boolean =
        { it.last().isUpperCase() }

private fun hasLengthLargerThen(length: Int): (String) -> Boolean =
        { it.length > length }

private fun hasLengthSmallerThen(length: Int): (String) -> Boolean =
        { it.length < length }

private fun contains(text: String): (String) -> Boolean =
        { it.contains(text) }

private val hasLengthInRange =
        hasLengthLargerThen(3).and(hasLengthSmallerThen(10))

private val hasCorrectFirstOrLastLetter =
        hasFirstCapitalLetter.or(hasLastCapitalLetter)

fun getStrings(data: List<String>) =
        Lists.select(
                hasLengthInRange.and(hasCorrectFirstOrLastLetter).and(contains("JUG")),
                data
        )
